src/sage/modules/free_module.py

```diff
@@ -6685,8 +6721,17 @@ def __init__(self, ambient, basis, check=True,
                 raise TypeError("each element of basis must be in "
                                 "the ambient vector space")
 
-        if echelonize and not already_echelonized:
-            basis = self._echelonized_basis(ambient, basis)
+        basis = basis_seq(V, basis)
+
+        # normalize the parameters. We have (basis, echelonize, echelonized_basis, already_echelonized)
+        if already_echelonized:
+            assert echelonized_basis is None or echelonized_basis == basis, "inconsistent parameters provided"
+            echelonized_basis = basis
+        if echelonize:
```

Comment:
```suggestion
        elif echelonize:
```
Let's not re-echelonize if it was already given and the flag was not set (e.g., the user forgot).

Reply:
this code change does nothing right?

* the only possible case where this matters is when both `already_echelonized` and `echelonize` is true.
* In that case,
    * before the change, the block above would make sure `echelonized_basis` is not `None`. So the part below is no-op.
    * after the change, the part below is no-op as well.

Reply:
That requires a lot more logical thinking than it should, so I think it's quite brittle. It also does a subtle change since it sets the basis to the echelonized basis (so it is not a no-op). In principle, the user might want the module to hold onto this, but I don't know if this will ever be used in practice.

Reply:
Suggestion applied, but see the comment below for discussion.